llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp

```diff
@@ -1760,16 +1760,42 @@ bool LoopVectorizationLegality::isVectorizableEarlyExitLoop() {
   assert(LatchBB->getUniquePredecessor() == SingleUncountableExitingBlock &&
          "Expected latch predecessor to be the early exiting block");
 
-  // TODO: Handle loops that may fault.
   Predicates.clear();
-  if (!isDereferenceableReadOnlyLoop(TheLoop, PSE.getSE(), DT, AC,
-                                     &Predicates)) {
+  SmallVector<LoadInst *, 4> NonDerefLoads;
+  bool HasSafeAccess =
+      TTI->supportsSpeculativeLoads()
+          ? isReadOnlyLoopWithSafeOrSpeculativeLoads(
```

Comment:
I think you can just always call `isReadOnlyLoopWithSafeOrSpeculativeLoads`, then do something like:

```
  if (!isReadOnlyLoopWithSafeOrSpeculativeLoads() || (NonDerefLoads.size() && !TTI->supportsSpeculativeLoads())) {
    report...
    return false;
  }
```

However, it's probably more complicated than that because I think you need to call `supportsSpeculativeLoads` for every speculative load using the specific loaded memory type.